gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/NewDagManager.java

```diff
@@ -0,0 +1,430 @@
+      String nextDagId = entry.getKey();
+      Set<Dag.DagNode<JobExecutionPlan>> dagNodes = entry.getValue();
+      for (Dag.DagNode<JobExecutionPlan> nextDagNode: dagNodes) {
+        this.dagManagementStateStore.addJobState(nextDagId, nextDagNode);
+      }
+    }
+
+    for (Dag.DagNode<JobExecutionPlan> dagNodeToClean: nodesToCleanUp) {
+      String dagId = DagManagerUtils.generateDagId(dagNodeToClean).toString();
+      this.dagManagementStateStore.deleteJobState(dagId, dagNodeToClean);
+    }
+  }
+
+  /**
+   * Method that defines the actions to be performed when a job finishes either successfully or with failure.
+   * This method updates the state of the dag and performs clean up actions as necessary.
+   */
+  public Map<String, Set<Dag.DagNode<JobExecutionPlan>>> onJobFinish(Dag.DagNode<JobExecutionPlan> dagNode)
+      throws IOException {
+    Dag<JobExecutionPlan> dag = this.dagManagementStateStore.getDagForJob(dagNode);
+    String dagId = DagManagerUtils.generateDagId(dag).toString();
+    String jobName = DagManagerUtils.getFullyQualifiedJobName(dagNode);
+    ExecutionStatus jobStatus = DagManagerUtils.getExecutionStatus(dagNode);
+    log.info("Job {} of Dag {} has finished with status {}", jobName, dagId, jobStatus.name());
+    // Only decrement counters and quota for jobs that actually ran on the executor, not from a GaaS side failure/skip event
+    if (this.userQuotaManager.releaseQuota(dagNode)) {
+      dagManagerMetrics.decrementRunningJobMetrics(dagNode);
+    }
+
+    switch (jobStatus) {
+      case FAILED:
+        dag.setMessage("Flow failed because job " + jobName + " failed");
+        dag.setFlowEvent(TimingEvent.FlowTimings.FLOW_FAILED);
+        dagManagerMetrics.incrementExecutorFailed(dagNode);
+        return Maps.newHashMap();
+      case CANCELLED:
+        dag.setFlowEvent(TimingEvent.FlowTimings.FLOW_CANCELLED);
+        return Maps.newHashMap();
+      case COMPLETE:
+        dagManagerMetrics.incrementExecutorSuccess(dagNode);
+        return submitNext(dagId);
+      default:
+        log.warn("It should not reach here. Job status is unexpected.");
+        return Maps.newHashMap();
+    }
+  }
+
+  /**
+   * Submit next set of Dag nodes in the Dag identified by the provided dagId
+   * @param dagId The dagId that should be processed.
+   * @return
+   * @throws IOException
+   */
+  private Map<String, Set<Dag.DagNode<JobExecutionPlan>>> submitNext(String dagId) throws IOException {
+    Dag<JobExecutionPlan> dag = this.dagManagementStateStore.getDag(dagId);
+    Set<Dag.DagNode<JobExecutionPlan>> nextNodes = DagManagerUtils.getNext(dag);
+    List<String> nextJobNames = new ArrayList<>();
+
+    //Submit jobs from the dag ready for execution.
+    for (Dag.DagNode<JobExecutionPlan> dagNode : nextNodes) {
+      this.dagTaskStream.addDagAction(DagManagerUtils.createDagActionFromDagNode(dagNode, DagActionStore.FlowActionType.ADVANCE));
```

Comment:
let's first solidify the rest of this PR (esp. around `DagMgmtStateStore`) and the let's revisit this Q:  do we need this `NewDM` intermediary, or could the `DagTaskStream` itself be the `EventBus` subscriber to the messages the KJSM sends?